metrics/src/builder.rs

```diff
@@ -0,0 +1,77 @@
+use crate::Receiver;
+use std::error::Error;
+use std::fmt;
+use std::time::Duration;
+
+/// Errors during receiver creation.
+#[derive(Debug, Clone)]
+pub enum BuilderError {
+    /// Failed to spawn the upkeep thread.
+    ///
+    /// As histograms are windowed, reads and writes require getting the current time so they can
+    /// perform the required maintenance, or upkeep, on the internal structures to roll over old
+    /// buckets, etc.
+    ///
+    /// Acquiring the current time is fast compared to most operations, but is a significant
+    /// portion of the other time it takes to write to a histogram, which limits overall throughput
+    /// under high load.
+    ///
+    /// We spin up a background thread, or the "upkeep thread", which updates a global time source
+    /// that the read and write operations exclusively rely on.  While this source is not as
+    /// up-to-date as the real clock, it is much faster to access.
+    UpkeepFailure,
+}
+
+impl Error for BuilderError {}
+
+impl fmt::Display for BuilderError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            BuilderError::UpkeepFailure => write!(f, "failed to spawn quanta upkeep thread"),
+        }
+    }
+}
+
+/// Builder for [`Receiver`].
+#[derive(Clone)]
+pub struct Builder {
+    pub(crate) histogram_window: Duration,
```

Comment:
do you actually need this to be pub at all?

Reply:
It's pub so that `Receiver` can access it when we call `Receiver::from_builder(builder_obj)`, but I'm open to suggestions.  I did start the process of creating `MetricsConfiguration` where those values are stored so that we can pass around a single struct instead of potentially multiple values needed when creating a metric.  Just not sure if `Builder` should generate that and pass it into `Receiver::from(..)`.

I'd be fine with making that change, although they'd still have to be either public fields or getters.

Thoughts?

Reply:
meh im fine with it that way

Reply:
I ended up creating a `Configuration` type that will hold all of this, and have the pub fields.  The `Builder` will just be the fluent API while `Configuration` can be cloned/copied/passed around for anything that needs the config data.